03_letters/shahriys/main_test.go

```diff
@@ -0,0 +1,52 @@
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"reflect"
+	"testing"
+)
+
+func TestSortLettersMain(t *testing.T) {
+	var bufletter bytes.Buffer
+	outletter = &bufletter
+
+	main()
+	expected := `a:2
+b:1`
+	actual := bufletter.String()
+
+	if expected != actual {
+		t.Errorf("Unexpected output in main()" + expected + " " + actual)
+	}
+
+}
+func TestSortLetters(t *testing.T) {
+	var bufletter bytes.Buffer
+	outletter = &bufletter
+
+	type test struct {
+		input    string
+		expected []string
+		actual   []string
+	}
+
+	tests := []test{
+		{input: "abac", expected: []string{"a:2", "b:1", "c:1"}, actual: sortLetters(letters("abac"))},
+		{input: "aAdbbDA", expected: []string{"A:2", "D:1", "a:1", "b:2", "d:1"}, actual: sortLetters(letters("aAdbbDA"))},
+		{input: "aAAADDDbac", expected: []string{"A:3", "D:3", "a:2", "b:1", "c:1"},
+			actual: sortLetters(letters("aAAADDDbac"))},
+		{input: "", expected: []string{}, actual: sortLetters(letters(""))},
+	}
+
+	for _, tc := range tests {
```

Comment:
I think you are on the right track here with creating a range of tests, but this can probably be cleaned up a little bit. Take a look at [Multiple Cases](https://goplay.space/#Zy2ta-z8vrO) for inspiration. A couple of points here:
 - I would generally avoid storing the `actual` value as a part of the test struct, there is no need to store it, rather generate it as the first step of the loop over the testCases.
 - I don't see the need to do the two `fmt.Printf()` calls when you are printing the expected and output as a part of the `t.Errorf()` call, please remove these.
 - If you use a `map[string]struct` you can use the key as a test name and easily structure your tests according to different cases, as seen in the example.

```go
cases := map[string]struct{input string, expected []string}{ "Test Name": {"input", {"expected"}} }

for name, test := range cases {
  t.Run(name, func(t *testing.T) {
    if actual := sortLetters(test.input); actual != test.expected {
      t.Errorf("error msg here")
    } 
  })
}

Reply:
done